0005-kzen-two-party-ecdsa.adoc

```diff
@@ -0,0 +1,173 @@
+The point of this spike is to review <<Lindell17>> and KZen's rust implementation <<kzen_mp_ecdsa>> and make notes about.
+
+
+== Overview
+
+<<Lindell17>> describes a two-party protocol to (i) generate a joint public key and (ii) jointly create ECDSA signatures under that public key.
+This is usually called signature _aggregation_.
+There's a video by the Lindell summarizing his paper here: https://www.youtube.com/watch?v=pwc_Ork-1aA
+
+Aggregated signatures are a nice thing to have.
+They look exactly like normal signatures but are produced by multiple parties instead of one.
+Practically, this means you don't need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`). There's no way to tell the difference between a single user `P2WSH` output and a `P2WSH` output from a joint public key.
+
+Since ECDSA signatures doesn't have a nice algebraic structure it is **very** difficult to do signature aggregation.
+In schnorr you can quite literally just add each party's share of the signature together.
+The method we have to use in ECDSA is to move the operations to produce the *s* component of the signature into a [Paillier group](https://en.wikipedia.org/wiki/Paillier_cryptosystem) rather than doing them simply modulo the Secp256k1 curve order.
+Once you're in Paillier land things become much easier but it comes at a cost of boatloads of messages and complexity during the key generation and exchange phase.
+It is so complicated that it's amazing that anyone actually bothered to come up with this.
+For us it may be very useful. 
+Even though Bitcoin will get Schnorr signatures, Ethereum probably won't switch anytime soon.
+
+The hard bit is doing the setup phase which involves quite a few proofs.
+Some of them are in the above crate but some of them are imported from elsewhere.
+We'll start with going through these.
+
+== Key Generation
+
+=== Generating the joint public key
+
+To produce signatures under a joint public key you first need to generate it fairly.
+Because of rogue key attacks, you can't just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing like protocol.
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key afterwards.
+
+Note, The ability to simulate a random coin toss gives the key generation very strong security guarantees (they are able to do a simulation based proof).
+
+The protocol is roughly as follows:
+
+- P1 -> Hash(p1_public_key, schnorr_nizk(p1_public_key))
+- P2 -> p2_public_key, schnorr_nizk(p2_public_key)
+- P1 -> Opens commitment from first message
+- Both parties calculate `joint_public_key = p1_public_key + p2_public_key`.
```

Comment:
I guess before that, both parties have to actually verify the ZKP?

Reply:
Yes they certainly must do that and p2 must verifying the opening of the commitment. I those steps out though because they seem obvious.